### ISSUE TITLE
Recurring jobs not starting anymore

### ISSUE THREAD

**Issue by voidmonk** [HangfireIO/Hangfire#1404]
Hi, I followed the [upgrade steps](https://docs.hangfire.io/en/latest/upgrade-guides/upgrading-to-hangfire-1.7.html#upgrade-steps) for latest `Hangfire` 1.7.x being used in a .NET Framework 4.7.2 Windows Service, by upgrading the `Hangfire.Core` NuGet package, and changing the configuration as follows:

```csharp
GlobalConfiguration.Configuration
    .SetDataCompatibilityLevel(CompatibilityLevel.Version_170)
    .UseSimpleAssemblyNameTypeSerializer()
    .UseStorage(...)
````

The job server starts, but recurring jobs are not picked up at all, although they appear in the dashboard with a status like: `Next execution: 14 minutes ago`.

I'm using Redis storage (via [Hangfire.Redis.StackExchange](https://github.com/marcoCasamento/Hangfire.Redis.StackExchange) package by @marcoCasamento). I've even done a `FLUSHALL` to wipe out all storage data. Also tried the `UseRecommendedSerializerSettings()` config option, but no luck.

It all worked fine with v1.6.x until today, so I'm guessing something trivial related to the upgrade is causing the issue in my case.

There's no evident error or runtime exception. Any guesses?

**Comment by odinserj** [comment]
Did you install 1.7.1 released yesterday or using 1.7.0? The latest version shows problems with invalid Cron expressions in Dashboard UI, so you can fix the problem.

Also, there should be error message in the log. But of course, you need to ensure logging is enabled for Hangfire.

**Comment by voidmonk** [comment]
I've narrowed down the problem in a job filter (duh).

It seems `CreatingContext`'s `InitialState` is null in 1.7.x, so this code that tries to find a job's queue name doesn't work:

```csharp
internal class HangfireSingletonFilter : JobFilterAttribute, IClientFilter
{
    public void OnCreating(CreatingContext context)
    {
        // context.InitialState is null in 1.7.x, but worked in 1.6.x
        var queueName = ((EnqueuedState)context.InitialState).Queue;
    }

    void IClientFilter.OnCreated(CreatedContext context)
    {
    }
}
```

Is there an alternative approach I can try?

**Comment by odinserj** [comment]
Oh development, fixing one problem and getting another 🙁 Where do you pass that queue name? As a quick suggestion you can use IElectStateFilter or IApplyStateFilter instead.

**Comment by voidmonk** [comment]
> Where do you pass that queue name?

Queue name is set initially in:

```csharp
RecurringJob.AddOrUpdate("TestWebJob", () => TestWebJob(JobCancellationToken.Null), 
    Cron.Minutely(), queue: "TestQueue");
```

And also applied as a method attribute:

```csharp
[Queue("TestQueue"), HangfireSingletonFilter, AutomaticRetry(Attempts = 3)]
public static void TestWebJob(IJobCancellationToken CancellationToken)
{
}
```

**Comment by odinserj** [comment]
And where is it passed? I.e. who use the queue name?

**Comment by voidmonk** [comment]
It gets passed to `EnqueuedCount()` and `EnqueuedJobs` methods of `context.Storage.GetMonitoringApi()`, to eventually cancel a job if it's already running (based on metrics).

Looks like your gist ([SkipWhenPreviousJobIsRunningAttribute](https://gist.github.com/odinserj/a6ad7ba6686076c9b9b2e03fcf6bf74e)) is meant for the same purpose - limit execution of job to one at a time (singleton). Is that gist code still up-to-date?

**Comment by voidmonk** [comment]
I tried your filter from the gist, and it works great (more efficient than the other one). Only change I had to make was replace `JobHelper.FromJson` with `SerializationHelper.Deserialize`.

So finally, (1) do we absolutely need `SetDataCompatibilityLevel(CompatibilityLevel.Version_170)` with a fresh/blank storage, and (2) should we use `UseSimpleAssemblyNameTypeSerializer()` or `UseRecommendedSerializerSettings()`?

Thank you.
